utils/zerovec/src/yoke_impls.rs

```diff
@@ -148,6 +149,96 @@ where
     }
 }
 
+/// This impl can be made available by enabling the optional `yoke` feature of the `zerovec` crate
+#[allow(clippy::transmute_ptr_to_ptr)]
+unsafe impl<'a, K0, K1, V> Yokeable<'a> for ZeroMap2d<'static, K0, K1, V>
+where
+    K0: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    K1: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    V: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    <K0 as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
+    <K1 as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
+    <V as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
+{
+    type Output = ZeroMap2d<'a, K0, K1, V>;
+    fn transform(&'a self) -> &'a Self::Output {
+        unsafe {
+            // Unfortunately, because K and V are generic, rustc is
+            // unaware that these are covariant types, and cannot perform this cast automatically.
+            // We transmute it instead, and enforce the lack of a lifetime with the `K0, K1, V: 'static` bound
+            mem::transmute::<&Self, &Self::Output>(self)
+        }
+    }
+    fn transform_owned(self) -> Self::Output {
+        debug_assert!(mem::size_of::<Self::Output>() == mem::size_of::<Self>());
+        unsafe {
+            // Similar problem as transform(), but we need to use ptr::read since
+            // the compiler isn't sure of the sizes
+            let ptr: *const Self::Output = (&self as *const Self).cast();
+            mem::forget(self);
+            ptr::read(ptr)
+        }
+    }
+    unsafe fn make(from: Self::Output) -> Self {
+        debug_assert!(mem::size_of::<Self::Output>() == mem::size_of::<Self>());
+        let ptr: *const Self = (&from as *const Self::Output).cast();
+        mem::forget(from);
+        ptr::read(ptr)
+    }
+
+    fn transform_mut<F>(&'a mut self, f: F)
+    where
+        F: 'static + for<'b> FnOnce(&'b mut Self::Output),
+    {
+        unsafe { f(mem::transmute::<&mut Self, &mut Self::Output>(self)) }
+    }
+}
+
+/// This impl can be made available by enabling the optional `yoke` feature of the `zerovec` crate
+#[allow(clippy::transmute_ptr_to_ptr)]
+unsafe impl<'a, K0, K1, V> Yokeable<'a> for ZeroMap2dBorrowed<'static, K0, K1, V>
+where
+    K0: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    K1: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    V: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
+    <K0 as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
+    <K1 as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
```

Comment:
hmm, this and the above ZeroMapBorrowed impl should probably be placing the bound on BorrowedVariant here

Reply:
What exactly is the bound you had in mind?

I tried this:

```
<<K0 as ZeroMapKV<'a>>::Container as ZeroVecLike<'a, K0>>::BorrowedVariant: for<'b> Yokeable<'b>,
```

But then I get

```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
   --> utils/zerovec/src/yoke_impls.rs:354:14
    |
354 |     #[derive(yoke::Yokeable)]
    |              ^^^^^^^^^^^^^^
```

because `ZeroVecLike<'a>` needs a lifetime parameter, but it's unclear which one I should use there.

Reply:
You'd have to use `'static` there I think.

Reply:
This works for me on main

```
diff --git a/utils/zerovec/src/yoke_impls.rs b/utils/zerovec/src/yoke_impls.rs
index 09232ae5..5fe9e97f 100644
--- a/utils/zerovec/src/yoke_impls.rs
+++ b/utils/zerovec/src/yoke_impls.rs
@@ -6,7 +6,7 @@
 #![allow(clippy::forget_copy)]
 
 use crate::map::ZeroMapBorrowed;
-use crate::map::ZeroMapKV;
+use crate::map::{ZeroMapKV, ZeroVecLike};
 use crate::ule::*;
 use crate::{VarZeroVec, ZeroMap, ZeroVec};
 use core::{mem, ptr};
@@ -111,8 +111,8 @@ unsafe impl<'a, K, V> Yokeable<'a> for ZeroMapBorrowed<'static, K, V>
 where
     K: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
     V: 'static + for<'b> ZeroMapKV<'b> + ?Sized,
-    <K as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
-    <V as ZeroMapKV<'static>>::Container: for<'b> Yokeable<'b>,
+    <<K as ZeroMapKV<'static>>::Container as ZeroVecLike<'static, K>>::BorrowedVariant: for<'b> Yokeable<'b>,
+    <<V as ZeroMapKV<'static>>::Container as ZeroVecLike<'static, V>>::BorrowedVariant: for<'b> Yokeable<'b>,
 {
     type Output = ZeroMapBorrowed<'a, K, V>;
     fn transform(&'a self) -> &'a Self::Output {
```